provider.go

```diff
@@ -1227,3 +1227,12 @@ func (e internalError) Error() string {
 }
 
 func (e internalError) Unwrap() error { return e.err }
+
+// GetTypeToken returns the type associated with the current call.
+func GetTypeToken(ctx context.Context) string {
```

Comment:
Should we offer this util function within `infer` that acts on `pulumi.Context` instead? We have only one use case for getting type tokens from context currently, and it might be easier for users to just do `infer.GetTypeToken(ctx)`.

Reply:
I have changed the signature so both `context.Context` and `*pulumi.Context` work: 
```go
func GetTypeToken[Ctx interface{ Value(any) any }](ctx Ctx) string
```